game/gbtd_gbr1.ksy

```diff
@@ -0,0 +1,297 @@
+  title: GBM file
+  license: MIT
+  ks-opaque-types: true
+  application: Gameboy Map Builder, GBMB
+  file-extension:
+    - gbm
+  tags:
+    - game
+
+doc: |
+  GBR1 files are used by Gameboy Map Builder, GBMB
+  to store maps created with the program.
+doc-ref: https://www.devrs.com/gb/hmgd/gbmb.html
+
+seq:
+  - id: magic
+    contents: GBO1
+  - id: objects
+    type: gbr1_object
+    repeat: eos
+
+types:
+  gbr1_object:
+    seq:
+      - id: marker
+        contents: HPJMTL
+      - id: object_type
+        type: u2
+        enum: type
+      - id: object_id
+        type: u2
+      - id: master_id
+        type: u2
+      - id: crc
+        type: u4
+      - id: object_length
+        type: u4
+      - id: body
+        size: object_length
+        type:
+          switch-on: object_type
+          cases:
+            type::producer:                   producer
+            type::map:                        map
+            type::map_tile_data:              map_tile_data
+            type::map_properties:             map_properties
+            type::map_property_data:          map_property_data
+            type::map_default_property_value: map_default_property_value
+            type::map_settings:               map_settings
+            type::map_property_colors:        map_property_colors
+            type::map_export_settings:        map_export_settings
+            type::map_export_properties:      map_export_properties
+
+    enums:
+      type:
+        0x0001: producer
+        0x0002: map
+        0x0003: map_tile_data
+        0x0004: map_properties
+        0x0005: map_property_data
+        0x0006: map_default_property_value
+        0x0007: map_settings
+        0x0008: map_property_colors
+        0x0009: map_export_settings
+        0x000A: map_export_properties
+
+    types:
+      producer:
+        seq:
+          - id: name
+            type: strz
+            size: 128
+          - id: version
+            type: strz
+            size: 10
+          - id: info
+            type: strz
+            size: 128
+
+      map:
+        seq:
+          - id: name
+            type: strz
+            size: 128
+          - id: width
+            type: u4
+          - id: height
+            type: u4
+          - id: prop_count
+            type: u4
+          - id: tile_file
+            type: strz
+            size: 256
+          - id: tile_count
+            type: u4
+          - id: prop_color_count
+            type: u4
+
+      map_tile_data:
+        types:
+          record:
+            seq:
+              - id: flipped_vertically
+                type: b1
+              - id: flipped_horizontally
+                type: b1
+              - id: reserved1
+                type: b3
+              - id: sgb_palette
+                type: b3
+              - id: reserved2
+                type: b1
+              - id: gbc_palette
+                type: b5
+              - id: tile_number
+                type: b10
```

Comment:
Does this format really use **little-endian** byte-sized integers and [**big-endian** bit-sized integers](https://doc.kaitai.io/user_guide.html#bit-ints-be)? That would be something quite rare and uncommon. The ["bit endianness"](https://doc.kaitai.io/user_guide.html#bit-endian) is usually the same as byte endianness in the entire format. That's because the bitfields are mostly designed to be packed into a byte-aligned integer, which is first parsed in the same endianness as other byte-aligned integers use, and then the value of such integer is unpacked into the individual bit field members (bit-sized integers) using bitwise AND `&` and SHR `>>` operations.

And reading multi-byte integers in a non-native endianness is mostly less efficient than in the native endianness (since [byte swapping](https://en.wikipedia.org/wiki/Endianness#Byte_swapping) has to be done), so that's something file format authors keep in mind and try to avoid that. Therefore mixing BE and LE integers in the same format does not make sense.

Reply:
I think it was a big mistake to introduce byte-swapping for it instead of relying on a bit ranges remaper.

Reply:
I'm not sure I completely follow here but the current implementation does produce the correct result. Running this test script: https://github.com/ZetaTwo/gameboy-project/blob/master/gbm-parser/tester.py with the test file in that repo gives the correct tile_number values:

```
...

Type: Type.map_tile_data, ID: 2
Size: 429
MapTileData
	Num records: 143
	MapTileData[0].tile_number = 0
	MapTileData[1].tile_number = 1
	MapTileData[2].tile_number = 2
	MapTileData[3].tile_number = 3
	MapTileData[4].tile_number = 4
	...
```

If they were the wrong endianness I would expect it to be something like: "0, 512, 256, 768, ...", right?